File: crates/lunarbase-math/src/slot0.rs
//! Bit-exact packing and decoding for the protocol's `Lane.slot0` word.

use crate::types::{MathError, U256};

const PRICE_BITS: usize = 112;
const FEE_BITS: usize = 20;
const THRESHOLD_BITS: usize = 7;
const BLOCK_BITS: usize = 40;
const BLOCK_DELAY_BITS: usize = 8;
const SLIPPAGE_K_BITS: usize = 32;
const RESERVED_BITS: usize = 14;

const PRICE_SHIFT: usize = 0;
const ASK_FEE_SHIFT: usize = 112;
const BID_FEE_SHIFT: usize = 132;
const THRESHOLD_SHIFT: usize = 152;
const THRESHOLD_ENABLED_SHIFT: usize = 159;
const UPDATE_BLOCK_SHIFT: usize = 160;
const EXISTS_SHIFT: usize = 200;
const PAUSED_SHIFT: usize = 201;
const BLOCK_DELAY_SHIFT: usize = 202;
const SLIPPAGE_K_SHIFT: usize = 210;
const RESERVED_SHIFT: usize = 242;

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
/// Native-width boundary view of the packed `Lane.slot0` word.
///
/// The hot path retains the raw `U256`; this view is used only for explicit
/// pack/unpack operations and therefore does not spread `U256` through fields
/// whose Solidity widths are known.
pub struct LaneSlot0 {
    /// WAD-denominated lane price stored in the low 112 bits.
    pub price: u128,
    /// Exact-in fee applied when converting cash into the lane asset.
    pub ask_fee_bps: u32,
    /// Exact-in fee applied when converting the lane asset into cash.
    pub bid_fee_bps: u32,
    /// Seven-bit price movement threshold copied from the Solidity layout.
    pub price_push_threshold: u8,
    /// Whether `price_push_threshold` participates in on-chain update policy.
    pub threshold_enabled: bool,
    /// EVM block at which the packed lane price was last updated.
    pub latest_update_block: u64,
    /// Whether Core currently exposes this lane.
    pub exists: bool,
    /// Whether swaps through this lane are disabled.
    pub paused: bool,
    /// Inclusive quote TTL in execution blocks after a price update.
    pub block_delay: u8,
    /// Lane-specific slippage coefficient in protocol BPS.
    pub slippage_k_bps: u32,
    /// Uninterpreted upper 14 bits preserved across decode/encode operations.
    pub reserved_high_bits: u16,
}

#[inline(always)]
fn field_mask(bits: usize) -> U256 {
    (U256::ONE << bits) - U256::ONE
}

#[inline(always)]
fn read_field(word: U256, shift: usize, bits: usize) -> U256 {
    (word >> shift) & field_mask(bits)
}

#[inline(always)]
fn write_field(word: U256, value: U256, shift: usize, bits: usize) -> U256 {
    let shifted_mask = field_mask(bits) << shift;
    (word & !shifted_mask) | ((value & field_mask(bits)) << shift)
}

fn validate_native(value: u128, bits: usize, field: &'static str) -> Result<(), MathError> {
    if bits < 128 && value >= (1u128 << bits) {
        return Err(MathError::FieldOverflow {
            field,
            bits: bits as u16,
        });
    }
    Ok(())
}

/// Decodes the exact Solidity storage layout into native-width fields.
pub fn decode_lane_slot0(word: U256) -> LaneSlot0 {
    LaneSlot0 {
        price: read_field(word, PRICE_SHIFT, PRICE_BITS)
            .try_into()
            .expect("112-bit price fits u128"),
        ask_fee_bps: read_field(word, ASK_FEE_SHIFT, FEE_BITS)
            .try_into()
            .expect("20-bit fee fits u32"),
        bid_fee_bps: read_field(word, BID_FEE_SHIFT, FEE_BITS)
            .try_into()
            .expect("20-bit fee fits u32"),
        price_push_threshold: read_field(word, THRESHOLD_SHIFT, THRESHOLD_BITS)
            .try_into()
            .expect("7-bit threshold fits u8"),
        threshold_enabled: read_field(word, THRESHOLD_ENABLED_SHIFT, 1) == U256::ONE,
        latest_update_block: read_field(word, UPDATE_BLOCK_SHIFT, BLOCK_BITS)
            .try_into()
            .expect("40-bit block fits u64"),
        exists: read_field(word, EXISTS_SHIFT, 1) == U256::ONE,
        paused: read_field(word, PAUSED_SHIFT, 1) == U256::ONE,
        block_delay: read_field(word, BLOCK_DELAY_SHIFT, BLOCK_DELAY_BITS)
            .try_into()
            .expect("8-bit block delay fits u8"),
        slippage_k_bps: read_field(word, SLIPPAGE_K_SHIFT, SLIPPAGE_K_BITS)
            .try_into()
            .expect("32-bit slippage K fits u32"),
        reserved_high_bits: read_field(word, RESERVED_SHIFT, RESERVED_BITS)
            .try_into()
            .expect("14-bit reserved field fits u16"),
    }
}

#[inline(always)]
/// Reads the 112-bit pushed price as a word ready for quote arithmetic.
pub fn lane_slot0_price(word: U256) -> U256 {
    read_field(word, PRICE_SHIFT, PRICE_BITS)
}

#[inline(always)]
/// Reads the 20-bit ask fee as a word ready for quote arithmetic.
pub fn lane_slot0_ask_fee_bps(word: U256) -> U256 {
    read_field(word, ASK_FEE_SHIFT, FEE_BITS)
}

#[inline(always)]
/// Reads the 20-bit bid fee as a word ready for quote arithmetic.
pub fn lane_slot0_bid_fee_bps(word: U256) -> U256 {
    read_field(word, BID_FEE_SHIFT, FEE_BITS)
}

#[inline(always)]
/// Reads the EVM block at which the lane price was last updated.
pub fn lane_slot0_latest_update_block(word: U256) -> u64 {
    read_field(word, UPDATE_BLOCK_SHIFT, BLOCK_BITS)
        .try_into()
        .expect("40-bit block fits u64")
}

#[inline(always)]
/// Reads whether the packed lane exists.
pub fn lane_slot0_exists(word: U256) -> bool {
    read_field(word, EXISTS_SHIFT, 1) == U256::ONE
}

#[inline(always)]
/// Reads whether the packed lane is paused.
pub fn lane_slot0_paused(word: U256) -> bool {
    read_field(word, PAUSED_SHIFT, 1) == U256::ONE
}

#[inline(always)]
/// Reads the packed inclusive quote TTL in execution blocks.
pub fn lane_slot0_block_delay(word: U256) -> u8 {
    read_field(word, BLOCK_DELAY_SHIFT, BLOCK_DELAY_BITS)
        .try_into()
        .expect("8-bit block delay fits u8")
}

#[inline(always)]
/// Reads the packed lane slippage coefficient.
pub fn lane_slot0_slippage_k_bps(word: U256) -> u32 {
    read_field(word, SLIPPAGE_K_SHIFT, SLIPPAGE_K_BITS)
        .try_into()
        .expect("32-bit slippage K fits u32")
}

#[inline(always)]
/// Replaces the packed existence bit.
pub fn set_lane_slot0_exists(word: U256, exists: bool) -> U256 {
    write_field(word, U256::from(exists), EXISTS_SHIFT, 1)
}

#[inline(always)]
/// Replaces the packed lane pause bit.
pub fn set_lane_slot0_paused(word: U256, paused: bool) -> U256 {
    write_field(word, U256::from(paused), PAUSED_SHIFT, 1)
}

/// Replaces the packed price-push threshold and its enable bit.
///
/// # Errors
///
/// Returns [`MathError::FieldOverflow`] when `price_push_threshold` exceeds
/// the contract's seven-bit field.
pub fn set_lane_slot0_price_push_threshold(
    word: U256,
    price_push_threshold: u8,
    enabled: bool,
) -> Result<U256, MathError> {
    validate_native(
        price_push_threshold.into(),
        THRESHOLD_BITS,
        "pricePushThreshold",
    )?;
    let word = write_field(
        word,
        U256::from(price_push_threshold),
        THRESHOLD_SHIFT,
        THRESHOLD_BITS,
    );
    Ok(write_field(
        word,
        U256::from(enabled),
        THRESHOLD_ENABLED_SHIFT,
        1,
    ))
}

#[inline(always)]
/// Replaces the packed block-delay field.
pub fn set_lane_slot0_block_delay(word: U256, block_delay: u8) -> U256 {
    write_field(
        word,
        U256::from(block_delay),
        BLOCK_DELAY_SHIFT,
        BLOCK_DELAY_BITS,
    )
}

#[inline(always)]
/// Replaces the packed slippage coefficient.
pub fn set_lane_slot0_slippage_k_bps(word: U256, slippage_k_bps: u32) -> U256 {
    write_field(
        word,
        U256::from(slippage_k_bps),
        SLIPPAGE_K_SHIFT,
        SLIPPAGE_K_BITS,
    )
}

/// Encodes a native-width view into the exact 256-bit storage word.
///
/// # Errors
///
/// Returns [`MathError::FieldOverflow`] when a native value exceeds its
/// narrower Solidity storage field.
pub fn encode_lane_slot0(fields: &LaneSlot0) -> Result<U256, MathError> {
    validate_native(fields.price, PRICE_BITS, "price")?;
    validate_native(fields.ask_fee_bps.into(), FEE_BITS, "askFeeBps")?;
    validate_native(fields.bid_fee_bps.into(), FEE_BITS, "bidFeeBps")?;
    validate_native(
        fields.price_push_threshold.into(),
        THRESHOLD_BITS,
        "pricePushThreshold",
    )?;
    validate_native(
        fields.latest_update_block.into(),
        BLOCK_BITS,
        "latestUpdateBlock",
    )?;
    validate_native(
        fields.reserved_high_bits.into(),
        RESERVED_BITS,
        "reservedHighBits",
    )?;
    let mut word = U256::from(fields.price)
        | (U256::from(fields.ask_fee_bps) << ASK_FEE_SHIFT)
        | (U256::from(fields.bid_fee_bps) << BID_FEE_SHIFT)
        | (U256::from(fields.price_push_threshold) << THRESHOLD_SHIFT);
    if fields.threshold_enabled {
        word |= U256::ONE << THRESHOLD_ENABLED_SHIFT;
    }
    word |= U256::from(fields.latest_update_block) << UPDATE_BLOCK_SHIFT;
    if fields.exists {
        word |= U256::ONE << EXISTS_SHIFT;
    }
    if fields.paused {
        word |= U256::ONE << PAUSED_SHIFT;
    }
    word |= U256::from(fields.block_delay) << BLOCK_DELAY_SHIFT;
    word |= U256::from(fields.slippage_k_bps) << SLIPPAGE_K_SHIFT;
    word |= U256::from(fields.reserved_high_bits) << RESERVED_SHIFT;
    Ok(word)
}

/// Packs two `uint20` fees into the contract's `uint40` calldata field.
///
/// # Errors
///
/// Returns [`MathError::FieldOverflow`] when either fee exceeds `uint20`.
pub fn encode_update_fees(ask_fee_bps: u32, bid_fee_bps: u32) -> Result<u64, MathError> {
    validate_native(ask_fee_bps.into(), FEE_BITS, "askFeeBps")?;
    validate_native(bid_fee_bps.into(), FEE_BITS, "bidFeeBps")?;
    Ok(u64::from(ask_fee_bps) | (u64::from(bid_fee_bps) << FEE_BITS))
}

/// Unpacks a contract `uint40` fee field into native fee widths.
///
/// # Errors
///
/// Returns [`MathError::FieldOverflow`] when bits above `uint40` are set.
fn decode_update_fees(fees: u64) -> Result<(u32, u32), MathError> {
    validate_native(fees.into(), 40, "fees")?;
    let mask = (1u64 << FEE_BITS) - 1;
    Ok(((fees & mask) as u32, (fees >> FEE_BITS) as u32))
}

/// Applies the Solidity operator update, including threshold-triggered pause.
///
/// # Errors
///
/// Returns [`MathError::FieldOverflow`] for values outside the corresponding
/// Solidity widths.
pub fn apply_lane_update_slot0(
    previous: U256,
    price: u128,
    fees: u64,
    block_number: u64,
) -> Result<U256, MathError> {
    validate_native(price, PRICE_BITS, "price")?;
    validate_native(fees.into(), 40, "fees")?;
    validate_native(block_number.into(), BLOCK_BITS, "blockNumber")?;
    let (ask_fee_bps, bid_fee_bps) = decode_update_fees(fees)?;
    let mut fields = decode_lane_slot0(previous);
    let delta = price.abs_diff(fields.price);
    let exceeds_threshold = fields.threshold_enabled
        && fields.price != 0
        && delta * 100 > fields.price * u128::from(fields.price_push_threshold);
    fields.price = price;
    fields.ask_fee_bps = ask_fee_bps;
    fields.bid_fee_bps = bid_fee_bps;
    fields.latest_update_block = block_number;
    if exceeds_threshold {
        fields.paused = true;
    }
    encode_lane_slot0(&fields)
}

File: packages/math/src/slot0.ts
import { assertU256, MathError, type Word } from "./constants.js";
import { decimalNumberToBigInt, type DecimalRounding } from "./decimal.js";
import type { LaneSlot0 } from "./types.js";

function fieldMask(bits: bigint): bigint {
  return (1n << bits) - 1n;
}
function readField(word: bigint, shift: bigint, bits: bigint): bigint {
  return (word >> shift) & fieldMask(bits);
}
const SLOT0_PRICE_MASK = (1n << 112n) - 1n;
const SLOT0_FEE_MASK = (1n << 20n) - 1n;
const SLOT0_BLOCK_MASK = (1n << 40n) - 1n;
const SLOT0_SLIPPAGE_K_MASK = (1n << 32n) - 1n;

/** Decimal model fields required to construct one lane price update. */
export interface LaneModelQuoteNumbers {
  /** CASH per one lane asset; use the model's `S` anchor price. */
  anchorPrice: number;
  /** Ask spread expressed in conventional basis points. */
  askSpreadBps: number;
  /** Bid spread expressed in conventional basis points. */
  bidSpreadBps: number;
  /** Decimal precision of the Core settlement asset. */
  cashDecimals: number;
  /** Decimal precision of the lane asset. */
  assetDecimals: number;
}

/** Integer fields from a model quote that are stored in `Lane.slot0`. */
export interface LaneSlot0QuoteFields {
  /** Decimal-adjusted WAD anchor price ready for the low `uint112`. */
  price: bigint;
  /** Ask spread in LunarBase protocol BPS, ready for `uint20`. */
  askFeeBps: bigint;
  /** Bid spread in LunarBase protocol BPS, ready for `uint20`. */
  bidFeeBps: bigint;
}

function validateField(value: bigint, bits: bigint, field: string): void {
  assertU256(value, field);
  if (value > fieldMask(bits))
    throw new MathError("FIELD_OVERFLOW", `${field} does not fit in ${bits} bits`, field, bits);
}

/**
 * Converts a model anchor price into the exact fixed-point value stored in
 * slot0.
 *
 * Floating-point multiplication is deliberately avoided. The canonical
 * decimal representation of `price` is scaled by
 * `10 ** (cashDecimals - assetDecimals) * WAD`, matching `Lanes.sol`. The
 * scaled result must be exact and fit the contract's `uint112` price field.
 */
export function lanePriceFromNumber(price: number, cashDecimals: number, assetDecimals: number): bigint {
  validateTokenDecimals(cashDecimals, "cashDecimals");
  validateTokenDecimals(assetDecimals, "assetDecimals");
  const scaleDecimals = 18 + cashDecimals - assetDecimals;
  if (scaleDecimals < 0)
    throw new RangeError("cashDecimals - assetDecimals produces a sub-integer LaneSlot0 price scale");
  const encoded = decimalNumberToBigInt(price, scaleDecimals, "exact");
  validateField(encoded, 112n, "price");
  return encoded;
}

/**
 * Converts conventional basis points into LunarBase protocol BPS.
 *
 * The protocol denominator is `1_000_000`, so one conventional basis point is
 * exactly 100 protocol units. A model spread can contain more than two
 * fractional bps digits; `rounding` determines how those excess digits map to
 * the integer `uint20` field.
 */
export function laneFeeBpsFromConventionalBps(spreadBps: number, rounding: DecimalRounding = "nearest"): bigint {
  const encoded = decimalNumberToBigInt(spreadBps, 2, rounding);
  validateField(encoded, 20n, "feeBps");
  return encoded;
}

/**
 * Decodes the quote-critical decimal fields produced by the pricing model.
 *
 * Pass model `S` as `anchorPrice` and the directional spreads as
 * `spreadAskBps` and `spreadBidBps`. Inputs must use conventional basis
 * points; Q24-encoded fee values are not accepted.
 */
export function modelQuoteToLaneSlot0Fields(
  quote: LaneModelQuoteNumbers,
  feeRounding: DecimalRounding = "nearest",
): LaneSlot0QuoteFields {
  return {
    price: lanePriceFromNumber(quote.anchorPrice, quote.cashDecimals, quote.assetDecimals),
    askFeeBps: laneFeeBpsFromConventionalBps(quote.askSpreadBps, feeRounding),
    bidFeeBps: laneFeeBpsFromConventionalBps(quote.bidSpreadBps, feeRounding),
  };
}

function validateTokenDecimals(value: number, field: string): void {
  if (!Number.isSafeInteger(value) || value < 0 || value > 0xff)
    throw new RangeError(`${field} must fit an ERC-20 uint8`);
}

/** Decodes every packed field from the 256-bit lane storage word. */
export function decodeLaneSlot0(word: Word): LaneSlot0 {
  assertU256(word, "slot0");
  return {
    price: readField(word, 0n, 112n),
    askFeeBps: readField(word, 112n, 20n),
    bidFeeBps: readField(word, 132n, 20n),
    pricePushThreshold: readField(word, 152n, 7n),
    thresholdEnabled: readField(word, 159n, 1n) === 1n,
    latestUpdateBlock: readField(word, 160n, 40n),
    exists: readField(word, 200n, 1n) === 1n,
    paused: readField(word, 201n, 1n) === 1n,
    blockDelay: Number(readField(word, 202n, 8n)),
    slippageKBps: Number(readField(word, 210n, 32n)),
    reservedHighBits: word >> 242n,
  };
}
/** Returns the packed 112-bit lane price. */
export const laneSlot0Price = (word: Word): bigint => assertU256(word, "slot0") & SLOT0_PRICE_MASK;
/** Returns the packed 20-bit ask fee in basis points. */
export const laneSlot0AskFeeBps = (word: Word): bigint => (assertU256(word, "slot0") >> 112n) & SLOT0_FEE_MASK;
/** Returns the packed 20-bit bid fee in basis points. */
export const laneSlot0BidFeeBps = (word: Word): bigint => (assertU256(word, "slot0") >> 132n) & SLOT0_FEE_MASK;
/** Returns the packed 40-bit latest-update block number. */
export const laneSlot0LatestUpdateBlock = (word: Word): bigint =>
  (assertU256(word, "slot0") >> 160n) & SLOT0_BLOCK_MASK;
/** Returns the packed lane existence bit. */
export const laneSlot0Exists = (word: Word): boolean => ((assertU256(word, "slot0") >> 200n) & 1n) === 1n;
/** Returns the packed lane pause bit. */
export const laneSlot0Paused = (word: Word): boolean => ((assertU256(word, "slot0") >> 201n) & 1n) === 1n;
/** Returns the packed inclusive quote TTL in execution blocks. */
export const laneSlot0BlockDelay = (word: Word): number => Number((assertU256(word, "slot0") >> 202n) & 0xffn);
/** Returns the packed lane slippage coefficient. */
export const laneSlot0SlippageKBps = (word: Word): number =>
  Number((assertU256(word, "slot0") >> 210n) & SLOT0_SLIPPAGE_K_MASK);
function replaceField(word: Word, value: bigint, shift: bigint, bits: bigint): Word {
  const mask = fieldMask(bits) << shift;
  return assertU256((word & ~mask) | ((value & fieldMask(bits)) << shift), "slot0");
}

/** Replaces the packed existence bit. */
export const setLaneSlot0Exists = (word: Word, exists: boolean): Word => replaceField(word, exists ? 1n : 0n, 200n, 1n);
/** Replaces the packed lane pause bit. */
export const setLaneSlot0Paused = (word: Word, paused: boolean): Word => replaceField(word, paused ? 1n : 0n, 201n, 1n);
/** Replaces the packed price-push threshold and its enable bit. */
export function setLaneSlot0PricePushThreshold(word: Word, pricePushThreshold: number, enabled: boolean): Word {
  if (!Number.isSafeInteger(pricePushThreshold) || pricePushThreshold < 0 || pricePushThreshold > 0x7f)
    throw new RangeError("pricePushThreshold does not fit uint7");
  const updated = replaceField(word, BigInt(pricePushThreshold), 152n, 7n);
  return replaceField(updated, enabled ? 1n : 0n, 159n, 1n);
}
/** Replaces the packed block delay. */
export const setLaneSlot0BlockDelay = (word: Word, blockDelay: number): Word => {
  if (!Number.isSafeInteger(blockDelay) || blockDelay < 0 || blockDelay > 0xff)
    throw new RangeError("blockDelay does not fit uint8");
  return replaceField(word, BigInt(blockDelay), 202n, 8n);
};
/** Replaces the packed slippage coefficient. */
export const setLaneSlot0SlippageKBps = (word: Word, slippageKBps: number): Word => {
  if (!Number.isSafeInteger(slippageKBps) || slippageKBps < 0 || slippageKBps > 0xffff_ffff)
    throw new RangeError("slippageKBps does not fit uint32");
  return replaceField(word, BigInt(slippageKBps), 210n, 32n);
};
/** Encodes a validated lane slot into the canonical bit layout. */
export function encodeLaneSlot0(fields: LaneSlot0): Word {
  validateField(fields.price, 112n, "price");
  validateField(fields.askFeeBps, 20n, "askFeeBps");
  validateField(fields.bidFeeBps, 20n, "bidFeeBps");
  validateField(fields.pricePushThreshold, 7n, "pricePushThreshold");
  validateField(fields.latestUpdateBlock, 40n, "latestUpdateBlock");
  if (!Number.isSafeInteger(fields.blockDelay) || fields.blockDelay < 0 || fields.blockDelay > 0xff)
    throw new RangeError("blockDelay does not fit uint8");
  if (!Number.isSafeInteger(fields.slippageKBps) || fields.slippageKBps < 0 || fields.slippageKBps > 0xffff_ffff)
    throw new RangeError("slippageKBps does not fit uint32");
  validateField(fields.reservedHighBits, 14n, "reservedHighBits");
  let word =
    fields.price | (fields.askFeeBps << 112n) | (fields.bidFeeBps << 132n) | (fields.pricePushThreshold << 152n);
  if (fields.thresholdEnabled) word |= 1n << 159n;
  word |= fields.latestUpdateBlock << 160n;
  if (fields.exists) word |= 1n << 200n;
  if (fields.paused) word |= 1n << 201n;
  word |= BigInt(fields.blockDelay) << 202n;
  word |= BigInt(fields.slippageKBps) << 210n;
  word |= fields.reservedHighBits << 242n;
  return assertU256(word, "slot0");
}
/** Packs ask and bid fees into the 40-bit update-fee payload. */
export function encodeUpdateFees(askFeeBps: bigint, bidFeeBps: bigint): bigint {
  validateField(askFeeBps, 20n, "askFeeBps");
  validateField(bidFeeBps, 20n, "bidFeeBps");
  return askFeeBps | (bidFeeBps << 20n);
}
/** Decodes the packed update-fee payload as `[askFeeBps, bidFeeBps]`. */
function decodeUpdateFees(fees: bigint): readonly [bigint, bigint] {
  validateField(fees, 40n, "fees");
  return [fees & fieldMask(20n), fees >> 20n];
}
/** Applies a price/fee/block update while preserving unrelated slot fields. */
export function applyLaneUpdateSlot0(previous: Word, price: bigint, fees: bigint, blockNumber: bigint): Word {
  validateField(price, 112n, "price");
  validateField(fees, 40n, "fees");
  validateField(blockNumber, 40n, "blockNumber");
  const [askFeeBps, bidFeeBps] = decodeUpdateFees(fees);
  const fields = decodeLaneSlot0(previous);
  const delta = price >= fields.price ? price - fields.price : fields.price - price;
  const exceedsThreshold =
    fields.thresholdEnabled && fields.price !== 0n && delta * 100n > fields.price * fields.pricePushThreshold;
  return encodeLaneSlot0({
    ...fields,
    price,
    askFeeBps,
    bidFeeBps,
    latestUpdateBlock: blockNumber,
    paused: exceedsThreshold || fields.paused,
  });
}
